vm-presets.md

```diff
@@ -0,0 +1,185 @@
+`VirtualMachine` is created, any applicable `VirtualMachinePresets`
+will be applied to the existing spec for the `VirtualMachine`. This allows
+for re-use of common settings that should apply to multiple `VirtualMachines`.
+
+
+Implementation
+------------------------
+
+`VirtualMachinePresets` are implemented as a Kubernetes `Initializer`. This
+means the `VirtualMachine` resource is modified before it is visible to or
+processed by any other component of KubeVirt.
+
+One noteworthy difference between the behavior of a `VirtualMachinePreset` and
+a `PodPreset` is that if an error occurs during processing of a `PodPreset`,
+Kubernetes will continue to create the `Pod` without applying the `PodPreset`.
+KubeVirt will move the VirtualMachine to the `Failed` state instead. The
+rationale behind this difference is that settings defined in a
+`VirtualMachinePreset` are presumed to be critical to the function of the
+`VirtualMachine`.
+
+Application of `VirtualMachinePresets` is all-or-nothing. If an error occurs
+while a `VirtualMachinePreset` is being applied (for example, a conflict
+occurs), none of the `VirtualMachinePreset` will be applied (and the
+`VirtualMachine` will be put into a `Failed` state, as noted above).
+
+Once a `VirtualMachinePreset` is successfully applied to a `VirtualMachine`,
+the `VirtualMachine` will be marked with an annotation to indicate that is
+was applied.
+
+Usage
+------------------------
+
+Any domain structure can be listed in the spec of a `VirtualMachinePreset`.
+e.g. Clock, Features, Memory, CPU, or Devices such network interfaces or disks.
+
+KubeVirt uses Kubernetes `Labels` to determine which `VirtualMachinePresets`
+apply to any given `VirtualMachine`, similarly to how `PodPresets` work in
+Kubernetes. The `VirtualMachine` is marked with Annotation upon successful
+completion.
+
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: testvm
+  selector:
+    matchLabels:
+      flavor: large-vm
+spec:
+  ...
+
+```
+
+Conflicts
+------------------------
+
+`VirtualMachinePresets` use a similar conflict resolution strategy to
+Kubernetes `PodPresets`. If a portion of the domain spec is present in both a
+`VirtualMachine` and a `VirtualMachinePreset` and both resources have the
+identical information, then no error will occur and `VirtualMachine` creation
+will continue normally. If however there is a conflict between the resources,
+an error will occur and the `VirtualMachine` will not be created. For example:
+If both the `VirtualMachine` and `VirtualMachinePreset` define a `Volume`, but
+use different paths for the same name, KubeVirt will note the conflict.
+
+
+Example
+=============================
+
+Here is an example of a `VirtualMachinePreset` and a `VirtualMachine` that
+uses it:
+
+
+```yaml
+kind: VirtualMachinePreset
+metadata:
+  name: windows-10
+  selector:
+    matchLabels:
+      flavor: windows-10
+spec:
```

Comment:
Could you synchornize the Preset examples with the current api? I think you see most of what you need in the windows-demo.yaml in the core repo.

Reply:
Done, thanks! I missed that example.